lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
     ch = ContextHandler = {
       contextPreviewExists: false
       UPDATE_DELAY: 1000
+      LABEL_SEQUENCE: [
+        {type: "keyword", value: /^\\label$/}
+        {type: "lparen", value: /^\{$/}
+        {type: "variable.parameter", value: /.*/}
+        {type: "rparen", value: /^\}$/}
+      ]
+      LABEL_PARAMETER_INDEX: 2
 
       getWholeEquation: (range) ->
-        editor.getSession().getTextRange(range)
+
+        tokenValues = []
+        labelSequenceIndex = 0
+        labelParameters = []
+        curLabelParameter = null
+        curLabelTokens = []
+
+        tokenIterator = new TokenIterator(editor.getSession(), range.start.row, range.start.column)
+        tokenIterator.stepForward() # at first it should be on the closing token of begin sequence
```

Comment:
you don't need to determine it. You just want the caller to pass properly positioned iterator. You already implicitly request to pass quite specific range: one which starts after \begin{equation} -- this is not visible in the function contract but you rely on it and do `stepForward` here with fingers crossed. I suggest moving this responsibility on the caller: let those who call the function position the iterator and do `stepForward` if needed.